src/compiler/common_sub_expression.cc

```diff
@@ -0,0 +1,63 @@
+/*!
+ *  Copyright (c) 2017 by Contributors
+ * \file common_sub_expression.cc
+ * \brief take common expressions out and replace with one.
+ *
+ */
+#include <nnvm/graph.h>
+#include <nnvm/pass.h>
+
+namespace nnvm {
+namespace compiler {
+
+nnvm::Graph CommonSubExpression(nnvm::Graph src) {
+  std::vector<NodeEntry> commonNodeList;
+  std::unordered_map<std::string, int> unique_expressn;
```

Comment:
I'm trying to understand this code. Seems you are mix-using `expression`, `expressn` and `exprssn`?

Reply:
Yes, please suggest one common term, will make it consistent.

In reply to your previous comment, CSE pass should apply only once in the input graph, it will modify the graph only if node exists which evaluates to same expressions, otherwise there wont be any change.

I hope this answer satisfis your query, please post if anything else. Thanks

Reply:
Thanks, @PariksheetPinjari909 . 
Personally, I would like to use `expr` or full `expression`.
For CSE here, I guess it's similar to the `merge` optimization in Theano. `merge` optimization will be applied several times since the computation graph may be changed by other optimizations. 
http://www.deeplearning.net/software/theano/optimizations.html#term-merge 
http://www.deeplearning.net/software/theano/extending/optimization.html#optdb-structure

Reply:
@TaoLv , i have gone through the link you shared, there also it is mentioned that the Merge can be present only once that is at the end.

And for the same reason i have placed the CSE pass after Optimize() and before precomputeprune() pass, as there wont be any Graph modification which will cause the Nodes which are different before and post evaluate to same. I feel once the Graph is filtered with CSE Pass, there is no need to run again.
Please let me know if i miss anything.